src/twisted/internet/tcp.py

```diff
@@ -1038,42 +1418,48 @@ def doRead(self):
                         # connection, but we get told to try to accept()
                         # anyway.
                         continue
-                    elif e.args[0] in (EMFILE, ENOBUFS, ENFILE, ENOMEM, ECONNABORTED):
-                        # Linux gives EMFILE when a process is not allowed to
-                        # allocate any more file descriptors.  *BSD and Win32
-                        # give (WSA)ENOBUFS.  Linux can also give ENFILE if the
-                        # system is out of inodes, or ENOMEM if there is
-                        # insufficient memory to allocate a new dentry.
-                        # ECONNABORTED is documented as possible on all
-                        # relevant platforms (Linux, Windows, macOS, and the
-                        # BSDs) but occurs only on the BSDs.  It occurs when a
-                        # client sends a FIN or RST after the server sends a
-                        # SYN|ACK but before application code calls accept(2).
-                        # On Linux, calling accept(2) on such a listener
-                        # returns a connection that fails as though the it were
-                        # terminated after being fully established.  This
-                        # appears to be an implementation choice (see
-                        # inet_accept in inet/ipv4/af_inet.c).  On macOS,
-                        # such a listener is not considered readable, so
-                        # accept(2) will never be called.  Calling accept(2) on
-                        # such a listener, however, does not return at all.
+                    elif e.args[0] == EMFILE and _reservedFD.available():
+                        # Linux and other UNIX-like operating systems
+                        # return EMFILE when a process has reached its
+                        # soft limit of file descriptors.
+                        log.msg(
+                            "EMFILE encountered;"
+                            " releasing reserved file descriptor.")
+                        _reservedFD.release()
+                        # Ensure that we have one more iteration
+                        # available after this.
+                        if i == (numAccepts - 1):
+                            numAccepts += 1
+                        continue
```

Comment:
This looks like it fixes the issue I raised last time.  I'm a bit concerned about the fragility of the fix (with respect to future code changes), though.  The test suite doesn't cover the case where this matters (and it would be tricky to make it and make sure it continues to hit the right case).